pkg/kubelet/cm/devicemanager/topology_hints.go

```diff
@@ -0,0 +1,139 @@
+// ensures the Device Manager is consulted when Topology Aware Hints for each
+// container are created.
+func (m *ManagerImpl) GetTopologyHints(pod v1.Pod, container v1.Container) map[string][]topologymanager.TopologyHint {
+	deviceHints := make(map[string][]topologymanager.TopologyHint)
+
+	for resourceObj, amountObj := range container.Resources.Limits {
+		resource := string(resourceObj)
+		amount := int(amountObj.Value())
+
+		if m.isDevicePluginResource(resource) {
+			klog.Infof("[devicemanager-topology] %v is a resource managed by device manager.", resource)
+
+			if aligned := m.deviceHasTopologyAlignment(resource); !aligned {
+				klog.Infof("[devicemanager-topology] Device does not have a topology preference")
+				deviceHints[resource] = nil
+				continue
+			}
+
+			available := m.getAvailableDevices(resource)
+			if available.Len() < amount {
+				klog.Infof("[devicemanager-topology] Requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, amount, available.Len())
+				deviceHints[resource] = []topologymanager.TopologyHint{}
+				continue
+			}
+
+			klog.Infof("[devicemanager-topology] Available devices for resource %v: %v", resource, available)
+
+			deviceHints[resource] = m.generateDeviceTopologyHints(resource, available, amount)
+		}
+	}
+
+	return deviceHints
+}
+
+func (m *ManagerImpl) deviceHasTopologyAlignment(resource string) bool {
+	// All devices must have Topology set for us to assume they care about alignment.
+	for device := range m.allDevices[resource] {
+		if m.allDevices[resource][device].Topology == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (m *ManagerImpl) getAvailableDevices(resource string) sets.String {
+	// Gets Devices in use.
+	m.updateAllocatedDevices(m.activePods())
+	devicesInUse := m.allocatedDevices[resource]
+
+	// Strip all devices in use from the list of healthy ones.
+	return m.healthyDevices[resource].Difference(devicesInUse)
+}
+
+func (m *ManagerImpl) iterateDeviceCombinations(devices sets.String, amount int, callback func(sets.String)) {
+	// Internal helper function to accumulate the combination before calling the callback.
+	var iterate func(devices, accum []string)
+	iterate = func(devices, accum []string) {
+		// Base case: we have a combination of size 'amount'.
+		if len(accum) == amount {
+			callback(sets.NewString(accum...))
+			return
+		}
+		for i := range devices {
+			iterate(devices[i+1:], append(accum, devices[i]))
+		}
+	}
+	iterate(devices.List(), []string{})
+}
+
+func (m *ManagerImpl) generateDeviceTopologyHints(resource string, devices sets.String, amount int) []topologymanager.TopologyHint {
+	// Initialize minAffinity to a full affinity mask.
+	minAffinity, _ := socketmask.NewSocketMask()
+	minAffinity.Fill()
+
+	// Iterate through all combinations of devices and build hints from them.
+	hints := []topologymanager.TopologyHint{}
+	m.iterateDeviceCombinations(devices, amount, func(combination sets.String) {
```

Comment:
This looks awfully slow. IIUC, computational complexity for this loop is O(amount *  nDevices! / (amount! * (nDevices - amount)!)).

Modern SRIOV capable NICs usually has 64+ VFs. So, if someone will try to allocate 32 VFs for a single container this loop will try to generate over 10^18 hints and will hang here forever until the memory exhausted.
Maybe it's better to group all devices into buckets by socket_id and only iterate over socket combinations similar to what cpu manager does?
Number of sockets also could be too big, but it's fairly low in most of modern platforms. Hope we could handle about 10^4 hints without significant issues. However, this, IMHO, should be reworked in the future too.

Reply:
Yes. You are right.

I have an alternative that I implemented for the CPUManager that is O(num_cpus * num_socket_combinations) which should be OK since the number of sockets tends to be low: #73920

I plan to port that over here before it gets merged.

Reply:
Thanks.

Reply:
Thanks!
One more issue here that also might be covered by switching to sockets combination is that we have unordered list of devices. That means that for mask (1,1) we could allocate few devices from node0 and a few devices from node1 even if we able to allocate all of them from the same node. Using socket based combinations we'll have devices actually sorted by the socket_id and will be able to allocate them as local to each other as possible.

Reply:
Order shouldn’t matter here b cause this is just about the hint generation to figure out what affinity different combinations of devices have. Tja allocation happens at the GetAffinity() call where the combined hint from multiple components is consulted and devices are allocated with alignment to the socket mask int the combined hint.

Either way though, the socket combination enumeration is the right strategy no matter what because we should have one possible hint for each socket combination and need to check it no matter what.

Reply:
> Order shouldn’t matter here b cause this is just about the hint generation to figure out what affinity different combinations of devices have. Tja allocation happens at the GetAffinity() call where the combined hint from multiple components is consulted and devices are allocated with alignment to the socket mask int the combined hint.

Yes, you're right that order doesn't matter here while generating hints, but it matters in devicesToAllocate() function which has almost same code part. And if the resulted topology hint will have more than one node, allocated devices could be randomly distributed among them due to unsorted nature of the device list.

> Either way though, the socket combination enumeration is the right strategy no matter what because we should have one possible hint for each socket combination and need to check it no matter what.

Sure.

Reply:
Makes sense. I’ll need to double check what I’m doing on the allocation side and I’ll fix it up appropriately.